Balanced truncation tutorial

pmli opened the pull request:
This PR adds a balanced truncation tutorial. Right now, it is in a draft stage, and feedback on things to add/remove/change is welcome. Compared to `notebooks/heat.ipynb`, I decided to add multiple inputs and outputs to make it more interesting. Also, I assembled the matrices "by hand" to show it is not necessary to use the discretization toolkit (also, I'm not sure how to do it nicely for a system with multiple inputs).

lbalicki commented:
I think the tutorial shows how to use balanced truncation and the related functionality very well. Here are some thoughts I had while going through the tutorial:
- In the "Discretized model" section, there could be a few words about what a model represents in pyMOR. E.g. one could say model objects represent discretized systems, refer to the pymor.models package for an overview of the available models and then say that for the current problem `LTIModel` is of interest.
- In the "Running balanced truncation" section something similar could be done for the reductor object. E.g. it could be mentioned that reductor objects encapsule a specific model reduction method and the corresponding properties (e.g. error bounds, gramians, ...).
- In the line `rom = bt.reduce(10)` it could also be mentioned that reduction is possible with respect to a desired tolerance.

pmli commented:
Thanks @lbalicki.

> * In the "Discretized model" section, there could be a few words about what a model represents in pyMOR. E.g. one could say model objects represent discretized systems, refer to the pymor.models package for an overview of the available models and then say that for the current problem `LTIModel` is of interest.

I've added a small remark on `LTIModel` and where it can be found. I feel a more detailed discussion would fit better in the Technical Overview (i.e., the Explanation/Discussion section according to the [4 documentations system](https://documentation.divio.com/), although I'm not sure if we had a final decision about following this system).

> * In the "Running balanced truncation" section something similar could be done for the reductor object. E.g. it could be mentioned that reductor objects encapsule a specific model reduction method and the corresponding properties (e.g. error bounds, gramians, ...).

Again, I've added a small remark, but would be for adding a more detailed discussion to the Technical Overview.

> * In the line `rom = bt.reduce(10)` it could also be mentioned that reduction is possible with respect to a desired tolerance.

Sounds good, I've added this.

renefritze commented:
I've just merged the switch from nbplots to jupyter-sphinx. The `code-links` directive should be replaced with both a `:jupyter-download:notebook:`tutorial02` and :jupyter-download:script:`tutorial02` and the `nbplot` directives with `jupyter-execute`.  You can see a full usage example at https://raw.githubusercontent.com/pymor/pymor/cpp_tutorial/docs/source/tutorial_external_solver.rst

pmli commented:
Thanks @renefritze, it looks nicer now.

sdrave commented on docs/source/tutorial02.rst:
```diff
@@ -0,0 +1,194 @@
+Heat equation
+-------------
+
+We consider the following heat equation over :math:`[0, 1]` with two inputs
```
```suggestion
We consider the following one-dimensional heat equation over :math:`(0, 1)` with two inputs
```
(domains are always open sets)

sdrave commented on docs/source/tutorial02.rst:
```diff
@@ -0,0 +1,194 @@
+        & t > 0.
+    \end{align}
+
+In the following, we will create a discretized model and reduce it using the
```
```suggestion
In the following, we will create a discretized |Model| and reduce it using the
```

sdrave commented on docs/source/tutorial02.rst:
```diff
@@ -0,0 +1,194 @@
+    import matplotlib.pyplot as plt
+    import numpy as np
+    import scipy.sparse as sps
+    from pymor.basic import LTIModel, BTReductor
```
I would prefer either `from pymor.basic import *` or import the objects from their proper locations.

pmli replied:
I went with the longer version.

sdrave commented on docs/source/tutorial02.rst:
```diff
@@ -0,0 +1,194 @@
+    from pymor.basic import LTIModel, BTReductor
+
+Next, we can assemble the matrices based on a centered finite difference
+approximation:
```
Of course, it would be nicer to user pyMOR's builtin discretizations for this, however, the point evaluations are not available at the moment.

To have some consistency with tutorial 1, maybe mention that we show here how to build an `LTIModel` from custom matrices instead of using a discretizer as in tutorial 1 (maybe mention `to_lti`?)

pmli replied:
I've added remarks in the previous paragraph.

sdrave commented on docs/source/tutorial02.rst:
```diff
@@ -0,0 +1,194 @@
+
+    print(fom)
+
+To visualize the behavior of the `fom`, we can draw its magnitude plot
```
add a comment what a 'magnitude plot' is supposed to be.

pmli replied:
Done

sdrave commented on docs/source/tutorial02.rst:
```diff
@@ -0,0 +1,194 @@
+    print(f'Relative H2 error:   {(fom - rom).h2_norm() / fom.h2_norm():.3e}')
+
+To compute the :math:`\mathcal{H}_\infty` norms, pyMOR uses the dense solver
+from Slycot.
```
Explain what these means for other `Operators` than dense `NumpyMatrixOperators`?

Mention, that `V` and `W` can also be obtained from `reductor`?

pmli replied:
Done

sdrave commented:
@renefritze, when I say `make docs` for this branch I get warnings like

```
WARNING: Cell printed to stderr:
00:00 L2ProductP1: Integrate the products of the shape functions on each element
```

and make reports an error. What am I doing wrong?

renefritze commented:
The warning is not in the last build log: https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/357865
The output itself is from an info logger. Do you have a defaults file (or something else) that might change the logging setup?

 > make reports an error

Can you be more precise?

renefritze commented:
Also sphinx and jupyter-sphinx versions could be different.

sdrave commented:
>The warning is not in the last build log: https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/357865
>The output itself is from an info logger. Do you have a defaults file (or something else) that might change the logging setup?

No. Maybe some jupyter widget installation issues, or is this independent from jupyter?

> >   make reports an error

> Can you be more precise?

```
copying static files... ... erledigt
copying extra files... erledigt
dumping search index in English (code: en)... erledigt
dumping object inventory... erledigt
build mit Problemen beendet, 10 warnings.
make[1]: *** [Makefile:47: html] Fehler 1
make[1]: Verzeichnis „/home/stephan/Repos/pymor/docs“ wird verlassen
make: *** [Makefile:71: docs] Fehler 2
```
I guess the error is due to the warnings.


> Also sphinx and jupyter-sphinx versions could be different.

```
Sphinx==3.0.1
jupyter-sphinx==0.2.4
```

renefritze commented:
> > The warning is not in the last build log: https://zivgitlab.uni-muenster.de/pymor/pymor/-/jobs/357865
> > The output itself is from an info logger. Do you have a defaults file (or something else) that might change the logging setup?
> 
> No. Maybe some jupyter widget installation issues, or is this independent from jupyter?

Do you have the logger capture box in the html output? Maybe the extension isn't loaded for you? https://github.com/pymor/pymor/blob/master/src/pymor/discretizers/builtin/gui/jupyter/__init__.py#L82

pmli commented:
> @renefritze, when I say `make docs` for this branch I get warnings like
> 
> ```
> WARNING: Cell printed to stderr:
> 00:00 L2ProductP1: Integrate the products of the shape functions on each element
> ```
> 
> and make reports an error. What am I doing wrong?

I have the same issues. Maybe this can be addressed in #982?

sdrave commented:
> Do you have the logger capture box in the html output? Maybe the extension isn't loaded for you? https://github.com/pymor/pymor/blob/master/src/pymor/discretizers/builtin/gui/jupyter/__init__.py#L82

Seems like the extension isn't loaded automatically. In a jupyter notebook, I only get the log output widget when I manually call the function you referenced.

renefritze commented:
The Jupyter Docker images get a custom kernel config https://github.com/pymor/docker-all/blob/master/jupyter/ipython_kernel_config.py

sdrave commented on docs/source/tutorial_bt.rst:
```diff
@@ -0,0 +1,202 @@
+Tutorial: Reducing a heat equation using balanced truncation
+============================================================
+
+:jupyter-download:notebook:`tutorial02`
+:jupyter-download:script:`tutorial02`
```
```suggestion
```

See below ..

sdrave commented on docs/source/tutorial_bt.rst:
```diff
@@ -0,0 +1,202 @@
+    print(f'Relative Hinf error: {(fom - rom).hinf_norm() / fom.hinf_norm():.3e}')
+    print(f'Relative H2 error:   {(fom - rom).h2_norm() / fom.h2_norm():.3e}')
+
+To compute the :math:`\mathcal{H}_\infty` norms, pyMOR uses the dense solver
+from Slycot, and therefore all of the operators are converted to dense matrices.
```
```suggestion
.. note::

    To compute the :math:`\mathcal{H}_\infty` norms, pyMOR uses the dense solver
    from Slycot, and therefore all of the operators have to be converted to dense matrices.
    For large systems, this may be very expensive.
    
Download the code: :jupyter-download:script:`tutorial_bt` :jupyter-download:notebook:`tutorial_bt`
```

pmli commented:
> It would be great if you also add a few sentences saying what balanced truncation actually is.

Sure, sounds good. I've added some text, let me know what you think.

sdrave commented:
> > It would be great if you also add a few sentences saying what balanced truncation actually is.
> 
> Sure, sounds good. I've added some text, let me know what you think.

Looks good. Maybe one more half-sentence about what the Gramians have to do with the projection? Currently no connection is made in the text except from saying that there is a connection.

pmli commented:
> Looks good. Maybe one more half-sentence about what the Gramians have to do with the projection? Currently no connection is made in the text except from saying that there is a connection.

Ok, good, at first I wasn't sure how much detail I should go into. For now, I just added a small remark about orthonormalization, to mention that V and W do not have to be exactly the columns from T and S.

sdrave commented:
@pmli, I think we can merge this now?

pmli commented:
@sdrave Yes, I think so, if the tests pass.
